### ISSUE TITLE
Limit the depth of proof search

### ISSUE THREAD

**Issue by Zhang-Liao** [coq-tactician/coq-tactician#59]
Statistics of the commit 97e26da3abeb008e5888d9c1b65b96c7cbb92bb9 shows the longest proof found in the benchmark contains 27 tactics. We can cut the node in the proof search with the depth larger than a suitable number, e.g., 35.

**Comment by LasseBlaauwbroek** [comment]
What would the purpose of this be? I don't immediately see any benefit of such a limit,

**Comment by Zhang-Liao** [comment]
> What would the purpose of this be? I don't immediately see any benefit of such a limit,

Perhaps I do not entirely understand the diagonal search. I think the proof search may go too deep in some path which unlikely leads to proof. If we can cut the too deep node, we can control the proof search in a depth that is likely to find proof. Thus, we can try more appropriate tactics.

**Comment by LasseBlaauwbroek** [comment]
I don't think that argument works. To illustrate, I can turn your argument around to create exactly the opposite issue of this one:

> Title: Limit the branching factor of proof search
>
> Statistics of commit `x` shows the proof with the widest proof search has a maximum branching factor of `y`. We can limit the maximum branching allowed to a suitable number, e.g., `y + epsilon`. This will allow us to search for longer proofs.

**Comment by LasseBlaauwbroek** [comment]
That being said, it is perfectly possible that the "angle" of the diagonal search is not set properly. This has never really been properly benchmarked. Changing the angle to either be steeper or less steep could have a nice effect.

The best solution is to make the angle of the diagonal depend on the confidence the model has in its tactics. The higher the confidence, the steeper the angle should be.

**Comment by Zhang-Liao** [comment]
> That being said, it is perfectly possible that the "angle" of the diagonal search is net set properly. This has never really been properly benchmarked. Changing the angle to either be steeper of less steep could have a nice effect.
> 
> The best solution is to make the angle of the diagonal depend on the confidence the model has in its tactics. The higher the confidence, the steeper the angle should be.

I agree we should try to build a confidence model in the future. The issue written here is just a preliminary step to the confidence model
